### ISSUE TITLE
[HOLD for payment 2024-06-28] [$250] Timezone in the ND app is not updated after moving to another timezone location

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#34035]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.22-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @quinthar 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1704159771205659

## Action Performed:
1. In ND app timezone is set to Automatically determine the location 
2. User moves from timezone A to timezone B
3. Check the timezone in ND in chrome web and in mobile and desktop time

## Expected Result:
Timezone is updated in ND to B

## Actual Result:
Timezone is not updated

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

![image](https://github.com/Expensify/App/assets/38435837/cd28ac63-7570-49fd-ae6b-7e5e0b60e624)

![image (1)](https://github.com/Expensify/App/assets/38435837/8a1a11a6-6c8e-405a-9f2c-811967819c16)

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01059d3aa439712d2e</li>
        <li>Upwork Job ID: 1752475553672241152</li>
        <li>Last Price Increase: 2024-05-29</li>
<li>Automatic offers: </li>
<ul>
<li>alitoshmatov | Reviewer | 102612346</li>
<li>dominictb | Contributor | 102612348</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @MitchExpensify</details>

**Comment by MitchExpensify** [comment]
Not sure how to test this without but going to try with a VPN

**Comment by MitchExpensify** [comment]
I took a little trip to Frankfurt and hard refreshed while there but the timezone did not automatically update in NewDot

<img width="1179" alt="image" src="https://github.com/Expensify/App/assets/36425901/49239f98-5005-449c-af85-94b37a2840e4">

**Comment by MitchExpensify** [comment]
It didn't update after signing out and back in either

**Comment by MitchExpensify** [comment]
Lastly, I am not sure if this fits into any wave or VIP. What do you think about it being in #vip-vsb @quinthar?

**Comment by MitchExpensify** [comment]
Asking here https://expensify.slack.com/archives/C066HJM2CAZ/p1705179873734529

**Comment by MitchExpensify** [comment]
Checking in again on a home for this: https://expensify.slack.com/archives/C066HJM2CAZ/p1705530053332199?thread_ts=1705179873.734529&cid=C066HJM2CAZ

**Comment by MitchExpensify** [comment]
It does seem to update after a hard refresh every time for me, would you still consider it a bug if it updates upon refresh @quinthar?

**Comment by MitchExpensify** [comment]
Friendly bump on the above question @quinthar 👀

**Comment by MitchExpensify** [comment]
Labelling internal as I assume this needs to be handled on the backend, please let me know if that is incorrect

**Comment by MitchExpensify** [comment]
I believe an internal volunteer will be recruited as part of the VIP project as a next step here

**Comment by alitoshmatov** [comment]
No proposals to review here. Internal engineer might work on this one

**Comment by MitchExpensify** [comment]
@quinthar will you be recruiting an internal volunteer for this seeing as its LOW on #vip-vsb?

**Comment by alitoshmatov** [comment]
No proposals to review

**Comment by MitchExpensify** [comment]
> No proposals to review

Applying the help wanted label would probably help us! 😆

**Comment by MitchExpensify** [comment]
Shared in #expenisfy-open-source

**Comment by erdsal4** [comment]
It seems like the only place where we update the automatic timezone is here:
https://github.com/Expensify/App/blob/af68510ad375bc99679908ce26dafead3a8b3dc1/src/libs/Navigation/AppNavigator/AuthScreens.tsx#L97-L108

I'm thinking that the Intl API is probably caching the timezone and returning the cached value even after the device's timezone changes

**Comment by nayabatir1** [comment]
Hi, I've check this and found this issue in chrome. 
On changing timezone you need to quit chrome and restart to get latest timezone. In other browser like safari it automatically detect new timezone.
I've attached video for reference. I'm using two browser chrome and safari.

Even if you add `updateAutomaticTimezone` in other places I don't think it will work, as chrome don't return latest timezone.

https://github.com/Expensify/App/assets/91016903/c7b808de-601e-44ff-95de-1115f94b302a

**Comment by nayabatir1** [comment]
Contributor details
Your Expensify account email: nayabatir1@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~017cd2e4e73324ec08

**Comment by jsdev2547** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

### What is the root cause of that problem?
The Intl.DateTimeFormat().resolvedOptions().timeZone does not get the correct timezone location.
https://github.com/Expensify/App/blob/af68510ad375bc99679908ce26dafead3a8b3dc1/src/pages/settings/Profile/TimezoneInitialPage.js#L40


### What changes do you think we should make in order to solve the problem?
To get the correct user timezone location, ip address location can be looked up. Following sites provide user timezone information based on ip address.
http://ip-api.com/json
https://ipinfo.io/

When the user has automatic timezone setup, api can be used to automatically fetch the timezone based on ipaddress and then update the timezone.

### What alternative solutions did you explore? (Optional)

**Comment by erdsal4** [comment]
Contributor details
Your Expensify account email: erdsal.4@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01fd13fe6d585708cd

**Comment by MitchExpensify** [comment]
Proposal here for you @alitoshmatov

**Comment by MitchExpensify** [comment]
Bumped @alitoshmatov 1:1

**Comment by alitoshmatov** [comment]
Reviewing

**Comment by alitoshmatov** [comment]
@jsdev2547 Thank you for your proposal. I don't think you are in right direction using third party api to determine timezone is overdoing.

**Comment by alitoshmatov** [comment]
@MitchExpensify After testing I don't think this issue is valid. If timezone changes new dot timezone also changes but it is effective only after browser is closed and opened again. Even if we wanted to instant result it is still dependent on browser which is only picking up timezone change after relaunching it. 
https://github.com/Expensify/App/assets/59907218/513ef3f7-5cfb-4a08-913e-a2a1e892d448

**Comment by MitchExpensify** [comment]
Are you saying that this is a browser issue, not a NewDot issue @alitoshmatov ? I.e. There is nothing we can do to avoid having to close and reopen the browser to see the timezone update?

**Comment by jsdev2547** [comment]
@alitoshmatov I have alternative approach without using third party api. The chrome browser issue is resolved in chrome version 124.0.6342.3 but the browser issue does not resolve the timezone issue. The timezone issue is occurring due to Intlpolyfil.

**Comment by alitoshmatov** [comment]
@MitchExpensify Yes

**Comment by alitoshmatov** [comment]
@MitchExpensify I believe so

**Comment by jsdev2547** [comment]
@alitoshmatov The browser issue is chrome specific and is resolved in Chrome dev 124.0.6342.3. The timezone issue is separate from browser issue and is due to using Intlpolyfil. Please refer to alternate [proposal](https://github.com/Expensify/App/issues/34035#issuecomment-1974760728).

**Comment by alitoshmatov** [comment]
> Alternatively, use polyfill instead of polyfill-force,
require('@formatjs/intl-datetimeformat/polyfill');
The polyfill calls shouldPolyfill to determine if the polyfill is needed.

@jsdev2547 I tried this but it is not working. I changed `    require('@formatjs/intl-datetimeformat/polyfill-force');
` to `    require('@formatjs/intl-datetimeformat/polyfill');
`. Any more steps which I missed

**Comment by alitoshmatov** [comment]
@jsdev2547 So as far as I know your solution to work we need chrome version 124 right?

**Comment by jsdev2547** [comment]
@alitoshmatov To test please download the chrome dev build and make sure the version of the build is greater than 124 or 124.0.6315.0. The chrome stable or release build is 122.0.6261.112 and still has the browser issue.

**Comment by jsdev2547** [comment]
@alitoshmatov The chrome dev build can be downloaded from the chrome release channels. When testing make sure to go in about chrome and check version is greater than 124 or 124.0.6315.0.

**Comment by alitoshmatov** [comment]
@jsdev2547 So that means it only works in dev version which our users are not using. The solution working only on our local build is not sufficient. We come back to my point which we cannot do anything right now because this issue is dependent on browser

**Comment by jsdev2547** [comment]
@alitoshmatov even if the chrome dev build is rolled out to chrome stable or release build the timezone issue would not get resolved. So the timezone issue would still need to get resolved. I suggest to use polyfill instead of polyfill-force and when the chrome dev channel is rolled out to chrome stable then the timezone would be automatically resolved and no need to wait for chrome dev to roll out to chrome stable.

**Comment by alitoshmatov** [comment]
Then I would suggest holding this issue, when needed chrome version releases we can come back to this issue

cc: @MitchExpensify

**Comment by jsdev2547** [comment]
@alitoshmatov The point I am trying to get across is the chrome browser issue and the timezone issue are separate. I am suggesting to use chrome dev build to test the timezone issue and not wait for the chrome stable version because the issue alone is not resolved based on the chrome browser issue and requires using polyfill.

**Comment by alitoshmatov** [comment]
I get what you mean. But I am saying that even if we solve this issue right now with chrome in dev version, I don't think it would be a good idea to merge a piece of code that relies on future releases of chrome. I suggest we hold this issue and when chrome launches its new version into production, then we proceed with solution.

I hope everything clear, if you still disagree with me, I think, we can ask internal engineers opinion on this matter

**Comment by MitchExpensify** [comment]
Holding for the right Chrome version makes sense to me. The version we're waiting for needs to be "greater than 124 or 124.0.6315.0", is that right?

**Comment by alitoshmatov** [comment]
Yes

**Comment by MitchExpensify** [comment]
Cool, placed on hold until then!

**Comment by MitchExpensify** [comment]
Current production version is 123.0.6312.59 - Still on hold

**Comment by MitchExpensify** [comment]
Current production version is 123.0.6312.107 - Still on hold

**Comment by jsdev2547** [comment]
@MitchExpensify The chrome production is 124.0.6367.62. The issue can be removed from hold.
@alitoshmatov I have tested again on chrome version 124.0.6367.62. I have revised proposal. Currently, after using polyFill or shouldPolyFill, the timezone is set to system timezone, but the user needs to click on the wallet tab and then click on profile or refresh the browser. I have revised proposal to add useEffect and on component mount call updateAutomaticTimezone, so the tab switching or browser refresh is not required.

**Comment by jsdev2547** [comment]
@MitchExpensify The chrome production is 124.0.6367.62. The issue can be removed from hold.
@alitoshmatov I have tested again on chrome version 124.0.6367.62. I have revised proposal. After using polyFill or shouldPolyFill, the timezone is set to system timezone, but the user needs to click on the wallet tab and then click on profile or refresh the browser. To set the timezone to system timezone, need to add useEffect and on component mount call updateAutomaticTimezone, so the tab switching or browser refresh is not required.

**Comment by MitchExpensify** [comment]
Off hold

**Comment by MitchExpensify** [comment]
Waiting for proposals, Melvin

**Comment by jsdev2547** [comment]
@MitchExpensify I have proposal ready for review, perhaps need to reach out to @alitoshmatov.

**Comment by jsdev2547** [comment]
@MitchExpensify I have proposal ready for review.

**Comment by jsdev2547** [comment]
@MitchExpensify Please reach out to the reviewer, the [proposal](https://github.com/Expensify/App/issues/34035#issuecomment-1974760728) is ready for review.

**Comment by MitchExpensify** [comment]
@alitoshmatov can you take a look at the proposal please?

**Comment by jsdev2547** [comment]
@alitoshmatov Please kindly review the proposal. Thanks.

**Comment by alitoshmatov** [comment]
Working on it

**Comment by alitoshmatov** [comment]
We can go with @jsdev2547 's [proposal](https://github.com/Expensify/App/issues/34035#issuecomment-1974760728) specifically alternative proposal. 

C+ reviewed 🎀 👀 🎀

**Comment by jasperhuangg** [comment]
Nice! I think the alternative solution is simpler and makes more sense to implement.

**Comment by MitchExpensify** [comment]
@jsdev2547 what's the latest here?

**Comment by jasperhuangg** [comment]
It looks like @jsdev2547 removed their assignment, I'm opening this back up for proposals.

**Comment by jasperhuangg** [comment]
@tienifr I see you added `@formatjs/intl-datetimeformat/polyfill-force` in https://github.com/Expensify/App/pull/31074, is there any reason why you decided to go with that instead of `@formatjs/intl-datetimeformat/polyfill` that we should be aware of before accepting this proposal?

**Comment by jasperhuangg** [comment]
cc @alitoshmatov ^

**Comment by alitoshmatov** [comment]
Asked @tienifr about above [comment](https://github.com/Expensify/App/pull/31074/files#r1623639817). Let's wait for their response

**Comment by jasperhuangg** [comment]
Cool, I think @suneox proposal checks out then. @alitoshmatov could you give it a look?

**Comment by jasperhuangg** [comment]
Got it, after some investigation I'm hesitant to remove the polyfill library, and feel better about @dominictb's proposal. 

Thank you for the investigation and discussion!

**Comment by MitchExpensify** [comment]
*Payment summary*:

- $250 @alitoshmatov requires payment [automatic offer](https://www.upwork.com/nx/wm/offer/102612346) (Reviewer)
- $250 @dominictb requires payment [automatic offer](https://www.upwork.com/nx/wm/offer/102612348) (Contributor)

Thoughts on the BZ steps, @alitoshmatov ?

**Comment by alitoshmatov** [comment]
- The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/App/pull/31074
- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: https://github.com/Expensify/App/pull/31074/files#r1623639817
- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: No need
- Determine if we should create a regression test for this bug.


**Regression Test Proposal**

Prerequisite: `Automatically determine your location.` is turned on in timezone section in ND 
1. Go to profile > Timezone
2.  In your devices native settings change your timezone
3. Make sure Timezone is updated to what you have changed to
_You may be required to go to different page of ND and coming back to Timezone page_

Do we agree 👍 or 👎

**Comment by jasperhuangg** [comment]
Not overdue, BZ steps are moving along 👍

**Comment by jasperhuangg** [comment]
Ahhh melvin pls 😆

**Comment by situchan** [comment]
As @MitchExpensify is the issue owner, he's the only person who can remove Overdue by commenting.

**Comment by MitchExpensify** [comment]
Paid and contract ended for @alitoshmatov 

I need you to accept this offer to complete your payment @dominictb https://www.upwork.com/nx/wm/offer/102612348

**Comment by MitchExpensify** [comment]
I'll issue the additional payment as a bonus if you can accept the existing offer, thanks @dominictb

**Comment by MitchExpensify** [comment]
Nice, paid and contract ended!
